program/src/withdraw.rs

```diff
@@ -58,66 +65,89 @@ pub fn handler(
 fn access_control(req: AccessControlRequest) -> Result<(), FundError> {
   let AccessControlRequest {
     program_id,
-    amount,
+    amount: _,
     fund_acc_info,
     withdraw_acc_info,
     vault_acc_info,
     vault_authority_acc_info,
   } = req;
-  // todo: check balance
 
   if !withdraw_acc_info.is_signer {
     return Err(FundErrorCode::Unauthorized)?;
   }
 
-  let _ = access_control::fund(fund_acc_info, program_id)?;
-  let _ = access_control::vault_join(
-    vault_acc_info,
-    vault_authority_acc_info,
-    fund_acc_info,
-    program_id,
-  )?;
+  {
+    let _ = access_control::fund(fund_acc_info, program_id)?;
+    let _ = access_control::vault_join(
+      vault_acc_info,
+      vault_authority_acc_info,
+      fund_acc_info,
+      program_id,
+    )?;
+  }
 
   let _ = access_control::withdraw(program_id, fund_acc_info, withdraw_acc_info);
 
+  info!("access control withdraw success")
+
   Ok(())
 }
 
 fn state_transistion(req: StateTransistionRequest) -> Result<(), FundError> {
   let StateTransistionRequest {
     fund_acc,
+    accounts,
     fund_acc_info,
     withdraw_acc_info,
     vault_acc_info,
     vault_authority_acc_info,
     token_program_acc_info,
+    nft_mint_acc_info,
+    nft_token_acc_info,
     amount,
   } = req;
 
-  fund_acc.deduct(amount);
-  // transfer from program account to owner of fund
-  info!("invoking token transfer");
-  let withdraw_instruction = transfer(
-    &ID,
-    vault_acc_info.key,
-    withdraw_acc_info.key,
-    &vault_authority_acc_info.key,
-    &[],
-    amount as u64,
-  )?;
+  {
```

Comment:
when withdrawal happens it should be assumed the fund is not open anymore. There should be another handler for this specific tx.